src/AbstractEventStateMachine.php

```diff
@@ -0,0 +1,144 @@
+use Dhii\State\Exception\CouldNotTransitionExceptionInterface;
+use Dhii\Util\String\StringableInterface as Stringable;
+use Exception;
+use Psr\EventManager\EventManagerInterface;
+
+/**
+ * Abstract functionality for event-based state machines.
+ *
+ * @since [*next-version*]
+ */
+abstract class AbstractEventStateMachine
+{
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    protected function _transition($transition)
+    {
+        $event     = $this->_getTransitionEvent($transition);
+        $exception = null;
+
+        try {
+            $this->_getEventManager()->trigger($event);
+        } catch (Exception $_ex) {
+            $exception = $_ex;
+        }
+
+        if ($event->isTransitionAborted()) {
+            throw $this->_createCouldNotTransitionException(
+                $this->__('Transition was aborted'),
+                null,
+                $exception,
+                $transition
+            );
+        }
+
+        $this->_setState($this->_getNewState($event));
+
+        if ($exception !== null) {
+            throw $this->_createStateMachineException(
+                $this->__('The triggered event threw an exception'),
+                null,
+                $exception
+            );
+        }
+    }
+
+    /**
+     * Retrieves the event manager associated with this instance.
+     *
+     * @since [*next-version*]
+     *
+     * @return EventManagerInterface The event manager instance.
+     */
+    abstract protected function _getEventManager();
+
+    /**
+     * Retrieves the transition event instance for a transition.
+     *
+     * @since [*next-version*]
+     *
+     * @param string|Stringable $transition The transition related to the transition event.
+     *
+     * @return TransitionEventInterface The transition event instance.
+     */
+    abstract protected function _getTransitionEvent($transition);
+
+    /**
+     * Retrieves the new state for a given event transition.
+     *
+     * @since [*next-version*]
+     *
+     * @param TransitionEventInterface $event The event.
+     *
+     * @return string|Stringable The new state.
+     */
+    abstract protected function _getNewState(TransitionEventInterface $event);
+
+    /**
+     * Sets the state for this instance.
+     *
+     * @since [*next-version*]
+     *
+     * @param string|Stringable|null $state The state, or null.
+     */
+    abstract protected function _setState($state);
+
+    /**
+     * Creates an state machine related exception.
+     *
+     * @since [*next-version*]
+     *
+     * @param string|Stringable|null $message  The error message, if any.
+     * @param int|null               $code     The error code, if any.
+     * @param Exception|null         $previous The previous exception for chaining, if any.
+     *
+     * @return CouldNotTransitionExceptionInterface The created exception.
+     */
+    abstract protected function _createStateMachineException(
```

Comment:
Not used in this class. Why is it here?

Reply:
Used in [`AbstractEventStateMachine#46`](https://github.com/RebelCode/event-state-machine/blob/864cbf673fe18f93ef97b775b8894f562c95c749/src/AbstractEventStateMachine.php#L46)

Reply:
My bad.